Add `Rotation2d`

Jondolf opened the pull request:
# Objective

Rotating vectors is a very common task. It is required for a variety of things both within Bevy itself and in many third party plugins, for example all over physics and collision detection, and for things like Bevy's bounding volumes and several gizmo implementations.

For 3D, we can do this using a `Quat`, but for 2D, we do not have a clear and efficient option. `Mat2` can be used for rotating vectors if created using `Mat2::from_angle`, but this is not obvious to many users, it doesn't have many rotation helpers, and the type does not give any guarantees that it represents a valid rotation.

We should have a proper type for 2D rotations. In addition to allowing for potential optimization, it would allow us to have a consistent and explicitly documented representation used throughout the engine, i.e. counterclockwise and in radians.

## Representation

The mathematical formula for rotating a 2D vector is the following:

```
new_x = x * cos - y * sin
new_y = x * sin + y * cos
```

Here, `sin` and `cos` are the sine and cosine of the rotation angle. Computing these every time when a vector needs to be rotated can be expensive, so the rotation shouldn't be just an `f32` angle. Instead, it is often more efficient to represent the rotation using the sine and cosine of the angle instead of storing the angle itself. This can be freely passed around and reused without unnecessary computations.

The two options are either a 2x2 rotation matrix or a unit complex number where the cosine is the real part and the sine is the imaginary part. These are equivalent for the most part, but the unit complex representation is a bit more memory efficient (two `f32`s instead of four), so I chose that. This is like Nalgebra's [`UnitComplex`](https://docs.rs/nalgebra/latest/nalgebra/geometry/type.UnitComplex.html) type, which can be used for the [`Rotation2`](https://docs.rs/nalgebra/latest/nalgebra/geometry/type.Rotation2.html) type.

## Implementation

Add a `Rotation2d` type represented as a unit complex number:

```rust
/// A counterclockwise 2D rotation in radians.
///
/// The rotation angle is wrapped to be within the `]-pi, pi]` range.
pub struct Rotation2d {
    /// The cosine of the rotation angle in radians.
    ///
    /// This is the real part of the unit complex number representing the rotation.
    pub cos: f32,
    /// The sine of the rotation angle in radians.
    ///
    /// This is the imaginary part of the unit complex number representing the rotation.
    pub sin: f32,
}
```

Using it is similar to using `Quat`, but in 2D:

```rust
let rotation = Rotation2d::radians(PI / 2.0);

// Rotate vector (also works on Direction2d!)
assert_eq!(rotation * Vec2::X, Vec2::Y);

// Get angle as degrees
assert_eq!(rotation.as_degrees(), 90.0);

// Getting sin and cos is free
let (sin, cos) = rotation.sin_cos();

// "Subtract" rotations
let rotation2 = Rotation2d::FRAC_PI_4; // there are constants!
let diff = rotation * rotation2.inverse();
assert_eq!(diff.as_radians(), PI / 4.0);

// This is equivalent to the above
assert_eq!(rotation2.angle_between(rotation), PI / 4.0);

// Lerp
let rotation1 = Rotation2d::IDENTITY;
let rotation2 = Rotation2d::FRAC_PI_2;
let result = rotation1.lerp(rotation2, 0.5);
assert_eq!(result.as_radians(), std::f32::consts::FRAC_PI_4);

// Slerp
let rotation1 = Rotation2d::FRAC_PI_4);
let rotation2 = Rotation2d::degrees(-180.0); // we can use degrees too!
let result = rotation1.slerp(rotation2, 1.0 / 3.0);
assert_eq!(result.as_radians(), std::f32::consts::FRAC_PI_2);
```

There's also a `From<f32>` implementation for `Rotation2d`, which means that methods can still accept radians as floats if the argument uses `impl Into<Rotation2d>`. This means that adding `Rotation2d` shouldn't even be a breaking change.

---

## Changelog

- Added `Rotation2d`
- Bounding volume methods now take an `impl Into<Rotation2d>`
- Gizmo methods with rotation now take an `impl Into<Rotation2d>`

## Future use cases

- Collision detection (a type like this is quite essential considering how common vector rotations are)
- `Transform` helpers (e.g. return a 2D rotation about the Z axis from a `Transform`)
- The rotation used for `Transform2d` (#8268)
- More gizmos, maybe meshes... everything in 2D that uses rotation

BorisBoutillier commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+
+/// A counterclockwise 2D rotation in radians.
+///
+/// The rotation angle is wrapped to be within the `]-pi, pi]` range.
```
I don't think so, this is a common way to describe inclusion or not of the range boundaries. (open or close range boundaries).
Here  `]-pi,pi]` explicit states all values withn `-pi` `pi`, excluding `-pi` itself, but including `pi`.

torsteingrindvik replied:
Ah ok, I'm used to `(-pi, pi]`, but if it's common to do it the way stated then TIL and nvm the suggestion.

NthTensor replied:
I had never seen this before. Strongly prefer standard interval notation here.

Jondolf replied:
Fwiw, Nalgebra also uses this: https://docs.rs/nalgebra/latest/nalgebra/geometry/type.UnitComplex.html#method.angle

NthTensor replied:
This isn't a hill I'm willing to die on, but I do think the engine should pick one or the other for consistency. In absence of a style guide I suspect the anglophone notation is already more prevalent in the engine docs, but I will have to check.

Jondolf replied:
Both notations are valid, but it looks like the parenthesis version is generally more common, yeah. The main benefit of the reverse square bracket is that it avoids misunderstandings where one might think e.g. `(x, y)` is a point or some other pair of objects `x` and `y` instead of an open interval. But that's not an issue here, so I'm fine changing it to `(-pi, pi]`.

Jondolf replied:
Changed it to the parenthesis notation

torsteingrindvik commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+
+use crate::prelude::{Direction2d, Mat2, Vec2};
+
+/// A counterclockwise 2D rotation in radians.
```
I'm assuming this means counterclockwise wrt an axis pointing into the screen, could it be clarified in the comment?

Jondolf replied:
It's not strictly constrained to "pointing into the screen", as I imagine this could be used for e.g. 2D physics in a 3D game (like a 2D minigame) or some other case where you embed 2D in 3D.

Essentially, this is just a counterclockwise rotation in a 2D plane in a right-handed Cartesian coordinate system, like this picture from Wikipedia:

![kuva](https://github.com/bevyengine/bevy/assets/57632562/527ef0b1-19d9-4c65-8dcb-810636e29930)

I'm not sure how I'd explain this clearly in a doc comment though without making it unnecessarily complicated for people. I feel like counterclockwise is pretty unambiguous in 2D, since a Z axis doesn't really even exist there. It's just the opposite of the direction in which the hands of a clock move when observing the clock from the front.

torsteingrindvik replied:
> It's just the opposite of the direction in which the hands of a clock move when observing the clock from the front.

It's something like this I would have hoped to see in the docs, because then it's even more clear.

NthTensor replied:
This seems acceptable in the 2d case. In 3D we would need to note handedness.

Jondolf replied:
Yeah, but that is basically the definition of "counterclockwise", so I felt like it was kind of unnecessary. I can add something like that though.

NthTensor replied:
Counter-clockwise seemed unambiguous to me.

Jondolf replied:
I agree; I'll leave it as is for the time being since I'm not sure what I'd put there that would add value

tguichaoua commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+    /// A counterclockwise rotation of π/3 radians.
+    pub const FRAC_PI_3: Self = Self {
+        cos: 0.5,
+        sin: 0.866_025_4,
```
This value should be a constante to avoid "[magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming))" in code.
Especially since this value appears multiple times.

NthTensor replied:
It's already in a constant in all these places, and it's pretty clear what its sin pi / ). This isn't what I think of as a magic number.

Jondolf replied:
The exact value is `sqrt(3) / 2`, but Rust doesn't have `SQRT_3` as a constant. I could add it or at least add a comment here though.

Jondolf replied:
I think I'll just leave it as is. Constants aren't ideal for trig function values since there'd have to be several constants with the same value because of symmetricity (e.g. `cos(pi/8) == sin(pi/3)`)

tguichaoua commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+
+    /// A counterclockwise rotation of π/6 radians.
+    pub const FRAC_PI_6: Self = Self {
+        cos: 0.866_025_4,
```
ditto

tguichaoua commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+
+    /// A counterclockwise rotation of π/8 radians.
+    pub const FRAC_PI_8: Self = Self {
+        cos: 0.923_879_5,
+        sin: 0.382_683_43,
```
ditto

tguichaoua commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+    }
+}
+
+impl From<f32> for Rotation2d {
+    /// Creates a [`Rotation2d`] from a counterclockwise angle in radians.
+    fn from(rotation: f32) -> Self {
+        Self::from_radians(rotation)
+    }
+}
```
I'm not very fan of this: `f32` doesn't necessarily represent an angle in radians.
I would prefer the user to explicitly use `from_radians` or `from_degrees` instead.

NthTensor replied:
I think this is mostly for backwards compatibility, to allow a you to still just pass on radians everywhere you could before. IMO that's a used feature.

Jondolf replied:
A while ago I opened some discussion on adding an `Angle` type or `Radians` and `Degrees` types: #11132. The overall consensus seemed to be that Bevy should use radians for rotations everywhere, and you should still be able to pass implicit floats too.

Not having this impl would also make this a breaking change, and a lot of APIs would be a bit longer.

IQuick143 replied:
I agree this is a desirable trade-off. Radians are very common in all sorts of angle math and demanding an explicit conversion everywhere would be too verbose.

I would however consider explaining radians in beginner-oriented documentation/examples, because those are the most likely to expect degree-based APIs.

tguichaoua commented on crates/bevy_gizmos/src/gizmos.rs:
```diff
@@ -554,11 +554,17 @@ impl<'w, 's, T: GizmoConfigGroup> Gizmos<'w, 's, T> {
+    pub fn rect_2d(
+        &mut self,
+        position: Vec2,
+        rotation: impl Into<Rotation2d>,
```
IMHO, using `impl Into<Rotation2d>` everywhere is a bit overkill.

Jondolf replied:
The only disadvantage is probably that it might make the function signature look slightly more complicated.

It does add a bit of context and explicitness though, since `Rotation2d` should always be counterclockwise and in radians. If we also added more methods that maybe return 2D rotations, you'd be able to pass them to methods directly instead of first converting to radians and then having the methods convert them back to `Rotation2d` anyway.

In my mind, `Rotation2d` is the 2D equivalent of a `Quat`, and should be used everywhere for rotation similarly. Taking an `impl Into<Rotation2d>` can be more efficient and flexible.

NthTensor replied:
Agree. Into<X> is a powerful abstraction that we should take full advantage of.

Jondolf commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,332 @@
+    #[inline]
+    pub fn new(radians: f32) -> Self {
+        Self::from_radians(radians)
+    }
```
I removed it for now. We could also remove the `from_` to make the constructors shorter:

```rust
// Before
let rotation1 = Rotation2d::from_radians(PI / 2.0);
let rotation2 = Rotation2d::from_degrees(45.0);

// After
let rotation1 = Rotation2d::radians(PI / 2.0);
let rotation2 = Rotation2d::degrees(45.0);
```

The getters would still be `as_radians` and `as_degrees`. No strong preference either way for me though

Jondolf replied:
I changed it to the shorter one for now, I think it might look a bit nicer. It should be pretty obvious that it's a constructor and not a getter considering that it takes an argument and there's already `as_radians` and `as_degrees`.

RobWalt commented on crates/bevy_math/src/rotation2d.rs:
````diff
@@ -0,0 +1,381 @@
+    /// ```
+    #[inline]
+    pub fn slerp(self, end: Self, s: f32) -> Self {
+        let delta = end * self.inverse();
+        self * Self::radians(delta.as_radians() * s)
````
I think we could reuse `angle_between` here

`self * Self::radians(self.angle_between(end) * s)`

Makes it much easier to understand. At least in my personal opinion

IQuick143 commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,380 @@
+        //       directly instead of first converting to radians.
+        // Same as `Quat::is_near_identity`
+        let threshold_angle = 0.002_847_144_6;
+        self.as_radians().abs() < threshold_angle
```
This should work if you just check that:
```suggestion
        cos > 0.0 && sin.abs() < threshold_angle.sin()
```
`threshold_angle.sin()` is of course just another constant.

Conceptually the idea is to cut out a rectangle out of the unit circle, such that only the desired arc remains.

(Please test this suggestion I'm not 100% sure it's entirely correct)

RobWalt commented on crates/bevy_math/src/rotation2d.rs:
````diff
@@ -0,0 +1,380 @@
+        }
+    }
+
+    /// Performs a linear interpolation between `self` and `rhs` based on
+    /// the value `s`.
+    ///
+    /// When `s` is `0.0`, the result will be equal to `self`.  When `s`
+    /// is `1.0`, the result will be equal to `rhs`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_math::Rotation2d;
+    /// #
+    /// let rot1 = Rotation2d::IDENTITY;
+    /// let rot2 = Rotation2d::radians(std::f32::consts::FRAC_PI_2);
+    ///
+    /// let result = rot1.lerp(rot2, 0.5);
+    ///
+    /// assert_eq!(result.as_radians(), std::f32::consts::FRAC_PI_4);
+    /// ```
````
Personally I think that the docs for `lerp` and `slerp` could highlight the difference between the two a bit more. 

The current example for `lerp` doesn't really help since the `a=0.5` value might be cool to get some well-known results ($\frac{\pi}{4}$), but on the other hand `slerp`-ing with the same `a` value would yield the same result. 

Here are some ideas how we could improve the situation:

- for both `lerp` and `slerp` use the same input rotations and `a` for a second example. Maybe make this example a bit more "extreme" where the angle between the input and output is close to $\pi$
- a small section listing some of the use cases for each method to further inform the user what this is for

Jondolf replied:
I made the examples use some slightly more interesting values, and added more tests for lerp and slerp.

I also improved the docs a bit, but I'm not quite sure how to describe lerp nicely; it kinda gives an ease-in-out effect, but the closer the angles are to being opposite (like 0 and pi), the more "aggressive" the interpolation is, where it essentially snaps from the start to the end almost immediately. I also tried explaining what it does mathematically, but I'm not 100% sure if it's entirely correct

RobWalt replied:
Looks better now! Thank you for spending a bit more time on it!

NthTensor commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,380 @@
+
+    /// Creates a [`Rotation2d`] from the sine and cosine of an angle in radians.
+    ///
+    /// The rotation is only valid if `sin.asin() == cos.acos()`.
```
Agreed. I think maybe epsilon = 10^-7.

Jondolf replied:
Thanks, fixed!

NthTensor commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -0,0 +1,380 @@
+
+    /// Rotates a [`Direction2d`] by a [`Rotation2d`].
+    fn mul(self, rhs: Direction2d) -> Self::Output {
+        Direction2d::new_unchecked(self * *rhs)
```
For what value of normalized?

Jondolf replied:
I'm guessing it could cause some drift, especially if the rotation gets denormalized (due to e.g. successive operations when running physics), but this is also an issue for Quat and Dir3. It's kinda unavoidable unless we normalize after every rotation, which could be expensive

Jondolf commented on crates/bevy_math/src/rotation2d.rs:
````diff
@@ -0,0 +1,380 @@
+    /// ```
+    #[inline]
+    pub fn lerp(self, end: Self, s: f32) -> Self {
+        Self::from_sin_cos(self.sin.lerp(end.sin, s), self.cos.lerp(end.cos, s))
````
I added methods for normalizing and getting the length/norm of the complex number stored in `Rotation2d`, and changed this `lerp` to normalize the result.

For the case of opposite angles and `s == 0.5`, I made it just return the original angle for now (this is documented). I'm not sure if this is the best approach; we could also

- Just return NaN/infinity for the rotation
- Return the midpoint between the angles (on which side though?)
- Panic
- Something else?

NthTensor commented on crates/bevy_math/src/rotation2d.rs:
````diff
@@ -0,0 +1,580 @@
+    /// assert_eq!(result2.as_degrees(), 67.5);
+    /// ```
+    #[inline]
+    pub fn lerp(self, end: Self, s: f32) -> Self {
+        Self {
+            sin: self.sin.lerp(end.sin, s),
+            cos: self.cos.lerp(end.cos, s),
+        }
+        .try_normalize()
+        // Fall back to the start rotation.
+        // This can happen when `self` and `end` are opposite angles and `s == 0.5`,
+        // because the resulting rotation would be zero, which cannot be normalized.
+        .unwrap_or(self)
+    }
````
Can we rename this nlerp, for normalized lerp? Docs should emphasize this is more efficient than slerp, and has similar output when the difference between rotations is small.

I think making the first point a tie-breaker seems acceptable. I don't expect this will be used much when the two rotations are more than pi/2 apart, let alone approaching pi.

Jondolf replied:
I restructured the docs a bit and added the things you mentioned.

I also renamed `lerp` to `nlerp` for now. You could argue that it's a bit confusing since `slerp` should also keep the rotation normalized, assuming the input is normalized. Unity also just calls it [`Lerp`](https://docs.unity3d.com/ScriptReference/Quaternion.Lerp.html), at least for quaternions. Nalgebra seems to have both [`lerp`](https://docs.rs/nalgebra/latest/nalgebra/geometry/type.UnitQuaternion.html#method.lerp) *and* [`nlerp`](https://docs.rs/nalgebra/latest/nalgebra/geometry/type.UnitQuaternion.html#method.nlerp) though. I'm fine with either name.

NthTensor replied:
Unity also uses [`nlerp`](https://docs.unity3d.com/Packages/com.unity.mathematics@1.2/api/Unity.Mathematics.math.nlerp.html) in some places.

NthTensor reviewed (approved):
Looks good to me now that lerp is fixed.

Jondolf commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -148,14 +226,21 @@ impl Rotation2d {
+    /// Uses a precision threshold of `1e-6`.
+    #[inline]
+    pub fn is_normalized(self) -> bool {
+        let length = self.sin.hypot(self.cos);
```
I changed this to be equivalent to Glam's `Quat::is_normalized` (which uses [`Vec4::is_normalized`](https://github.com/bitshifter/glam-rs/blob/9899f6e40277d5b186b7a2350046072534197469/src/f32/scalar/vec4.rs#L587)).

Jondolf commented:
Addressed the review suggestions. I also updated the 2D bounding volume transformations to use `Rotation2d`, as discussed here: https://github.com/bevyengine/bevy/pull/11681#discussion_r1477142018

IQuick143 commented on crates/bevy_math/src/rotation2d.rs:
```diff
@@ -228,11 +227,10 @@ impl Rotation2d {
     pub fn is_normalized(self) -> bool {
-        let length = self.sin.hypot(self.cos);
-        length - 1.0 <= 1e-6
+        (self.length_squared() - 1.0).abs() <= 2e-4
```
Maybe a comment explaining why we use `2e-4` in error analysis might be good for future reference.
